src/react/providers/chat-room-provider.tsx

```diff
@@ -77,42 +128,116 @@ export const ChatRoomProvider: React.FC<ChatRoomProviderProps> = ({
 }) => {
   const client = useChatClient();
   const logger = useLogger();
-  logger.trace(`ChatRoomProvider();`, { roomId, options, release, attach });
+  logger.debug(`ChatRoomProvider();`, { roomId, options, release, attach });
+
+  // Set the initial room promise, we do this in a function to avoid rooms.get being called
+  // every time the component re-renders
+  // In StrictMode this will be called twice one after the other, but that's ok
+  const [value, setValue] = useState<ChatRoomContextType>(() => {
+    logger.debug(`ChatRoomProvider(); initializing value`, { roomId, options });
+    return { room: client.rooms.get(roomId, options).catch(), roomId: roomId, options: options };
+  });
 
-  const [value, setValue] = useState({ room: client.rooms.get(roomId, options) });
+  // Create an effect that changes the room when the id or options change
+  const prevId = useRef(roomId);
+  const prevOptions = useRef(options);
+  const prevClient = useRef(client);
+  const roomReleaseQueue = useRef(new RoomReleaseQueue(logger));
 
+  // update the release queue if the logger changes - as it means we have a new client
+  // and only if it actually changes, not because strict mode ran it twice
   useEffect(() => {
-    const room = client.rooms.get(roomId, options);
+    // Don't create a new queue if the logger hasn't actually changed
+    if (roomReleaseQueue.current.logger === logger) {
+      return;
+    }
+
+    logger.debug(`ChatRoomProvider(); updating release queue`);
+    roomReleaseQueue.current = new RoomReleaseQueue(logger);
+  }, [logger]);
 
-    // Update state if room instance has changed.
-    setValue((prev) => {
-      if (prev.room === room) {
-        return prev;
+  useEffect(() => {
```

Comment:
This is more of a nitpick, but would this become a bit simpler to reason about if all the attach-detach and get-release stay in the same useEffect? Even the queued up stuff.

I get the separation now after spending lots of time on this - one's responsible for detecting changes in room and do a release+get ON MOUNT (_has this changed since my last mount?_), and one is responsible for clearing up if the component is unmounted (and there won't be another mount, hence the need for the queue).

Maybe a few comments on top of the useEffect if not refactor?